### ISSUE TITLE
Implement widnow state sychronization between Qt and PM

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/qtbase-os2#80]
The OS/2 platform plugin (see #77) currently has very limited awareness of the state of the PM window. It basically only tracks the size/position changes made through PM (e.g. with the mouse). Things that need to be added:
* Changing size/position from Qt.
* Tracking window state changes (min/max/hidden/shown etc).
* Keyboard focus traversal.

**Comment by dmik** [comment]
`QPlatformWindow` docs state the following regarding window positioning in platform plugins:
> \li Window positioning: When calling \c{QWindow::setFramePosition()}, the flag
            \c{QWindowPrivate::positionPolicy} is set to \c{QWindowPrivate::WindowFrameInclusive}.
            This means the position includes the window frame, whose size is at this point
            unknown and the geometry's topleft point is the position of the window frame.

**Comment by dmik** [comment]
I've implemented things like `setWIndowTitle`, `raise/lower` and some others and there is more to come. I'm checking other plugins ATM to detect what is a must and what may be postponed.

**Comment by dmik** [comment]
I willl be also tracking various paint issues here for now as they might be related to state out-of-sync and lack of other basic data exchange between the platform plugin and PM. Current issues:

1. `examples/qpa/qrasterwindow` does not start painting until the window is moved.
2. `examples/richtext/textedit` (and many others): dismissing the combobox leaves the combobox button sunken. The same happens with the main menu bar's entries (dismissing the menu leaves the menu bar's entry sunken). Invalidating the respective window part causes it to properly redraw it. Somehow the invalidation made by Qt is lost it seems.

**Comment by dmik** [comment]
Commit cab60ac fixes 1 and commit 65a827f fixes 2. Size and position changing is done in f9ad511. And there are a few more things to do (some cpoied from the description to give them a number):
3. Make QWindow::setFramePosition work correctly (`qpa/windows` reveals this).
4. Track window state changes (minimized/maximized). Note that active/visible states are already tracked.
5. Make keyboard focus traversal work (this depends on #79).

**Comment by dmik** [comment]
Note that we will also address modality support within this ticket. I've already half way through it (among other things). This might require a `QPlatformDialogHelper` implementation to properly support it. This class provides much finer control over dialog functionality. In particular, it allows to enter the native dialog exec loop (`WinProcessDlg`) which will make modal dialogs behave exactly like native OS/2 modal dialogs (e.g. not allow them to be moved behind the dialog window etc).

**Comment by dmik** [comment]
The main problem with modality support is that besides window modality, there is also application modality in Qt which is missing in PM itself. If the window is application modal, it should not only block its direct "owner" window but also all other top-level windows of the application. PM natively only supports the latter. Support for the former should be added manually and it's rather hackish in Qt 4 — I intercept WM_ADJUSTWINDOWPOS there and basically disable SWP_ACTIVATE requests by redirecting them to the current modal window. It works half way in Qt 5 so far since internal handling of modal widgets is changed there a lot (became much more clean in general).

**Comment by dmik** [comment]
Window modality is done. The code is much simpler and cleaner than in Qt4 and appears to work function wise just the same — at least, according to the `manual/windowmodality` test.

Note that as opposed to the native PM behavior, blocked (and thus disabled) top-level windows **can** be moved and resized through an external interface (e.g. via the corresponding XCenter context menus for their window list buttons). They can't be made active though and can't be placed above the modal window (thier activation is always forwarded to the modal). However, this is just as it was in Qt 4 and it looks fine for me — might be even useful in some cases. Other platforms (e.g. macOS) allow to move blocked windows (but not be activated and change their z-order) as well.

**Comment by dmik** [comment]
I'm working on min/max/fullscreen states now and just to inform you — it's a total mess. There are way too many uncertainties about that from the Qt side, like:
- These states are binary flags. So they can be combined. However, there is no defined order of precedence for them and it's not clear how to proceed when they are combined.
- It's not defined what should happen if a window is moved/resized while in one of min/max/fullscreen states.
- ...

All this makes it really hard to create a proper implementation. It was a mess in Qt 4 as well (and took a long time to work it more or less).

**Comment by dmik** [comment]
Moreover, the `manual/windowflags` test is broken in some places and seems not to be maintained any more, along with other tests like `manual/windowgeometry`. Interestingly enough, these tests seem to have been fixed in Qt 5.12, but we are still at Qt 5.11...

**Comment by dmik** [comment]
Another uncertain thing is the minimized state. In PM, you can define what minimize does in WPS settings. It can be one of three modes: hide the window, minimize it to the minimized window viewer, minimize it to the desktop. The last two modes actually resize the window to a special size of 48x50 pixels and move them accordingly. If it is minimized to the viewer which is hidden (this is the default), then PM moves such a minimized window to a weird position to keep it off-screen (-32000, -32000).

Currently, Qt sees such resizes/moves as window geometry changes and will even paint in the minimized state (unless it is set up to hide the window). This is a potential source of problems because no Qt application will ever expect that. In Qt 4, we made the minimized window appear hidden in Qt terms and ignore any move/reisize events in minimized state to not confuse it. All Qt geometry changes landed in the special PM geometry restore zone so that if the window is un-minimized, it would get the right size. I will do the same in Qt 5.

**Comment by dmik** [comment]
I finally made it all work at an acceptable level. This is which assertions the current implementation does (in brief, from `QOS2Window::setWindowState`):
```
{
    // Window state change logic is as follows, in a lousy attempt to match Qt docs (which are very
    // uncertain in this regard) and possibly other platforms (that may differ):
    //
    // 1. Minimized state overrides everything (but can coexist with any state).
    // 2. FullScreen overrides Maximized (but can coexist with it).
    // 3. Maximized (obviously) overrides normal state (and cancels Fullscreen when comes from PM).
    // 4. Removing Minimized restores to what's left (FullScreen, Maximized or normal).
    // 5. Removing FullScreen restores to what's left (Maximized or normal) unless there is also Minimized.
    // 6. Removing Maximized restores to normal unless there is also Minimized or FullScreen.
    // 7. Restoring using PM sysmenu always restores to normal state.
    // 8. Moving/resizing a Maximized/FullScreen window doesn't cancel these states, and vice versa
    //    (this is how PM behaves as opposed e.g. to macOS).
    // 9. Moving a Minimized window programmatically from Qt doesn't change its actual position but
    //    changes its normal geometry (for later restoring).
```

More than enough for this ticket. Further work will be done in other tickets (e.g. #91).
